src/main/scala/is/hail/expr/types/TArray.scala

```diff
@@ -44,7 +44,7 @@ final case class TArray(elementType: Type, override val required: Boolean = fals
   def _typeCheck(a: Any): Boolean = a.isInstanceOf[IndexedSeq[_]] &&
     a.asInstanceOf[IndexedSeq[_]].forall(elementType.typeCheck)
 
-  override def str(a: Annotation): String = JsonMethods.compact(toJSON(a))
+  override def strNonNull(a: Annotation, config: PrintConfig): String = JsonMethods.compact(toJSON(a))
```

Comment:
We can't rely on the JSON representation of arrays any more because something the numbers inside: `[1.0, 2.0]` won't be formatted correctly. We should just construct the correct string ourselves. Something like:

```scala
"[" + a.asInstanceOf[IndexedSeq[Annotation]].map(elementType.str _).mkString(",") + "]"
```

Reply:
FWIW, people like to make these kinds of parameters `implicit` which avoids passing them around explicitly.

Reply:
can you explain some of that design perspective? It makes sense to me to pass it around explicitly here, and I worry that passing it implicitly might make it easy to use the wrong config accidentally.

Reply:
The implicit resolution should fail if there are two in scope at once, e.g.:
```
dking@wmb16-359 # scala
implicit Welcome to Scala 2.12.0 (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_101).
Type in expressions for evaluation. Or try :help.
scala> implicit val foo = 3
foo: Int = 3

scala> implicit val bar = 5
bar: Int = 5

scala> implicitly[Int]
<console>:14: error: ambiguous implicit values:
 both value foo of type => Int
 and value bar of type => Int
 match expected type Int
       implicitly[Int]
```

That said, with this design strategy, you have to be careful about when and where you load implicit parameters into scope. For example, importing an implicit `PrintConfig` at the top-level would create ambiguity everywhere.

---

The benefit of this approach shines more when you have multiple calls that need a PrintConfig. It avoids repeating the PrintConfig value over and over, thus making the code noisy.

I suspect, in this case it doesn't provide much benefit over the explicit passing approach we are currently using.

Reply:
It's something for us to keep if we start to pass PrintConfig around everywhere and starts to get noisy.

Reply:
> The benefit of this approach shines more when you have multiple calls that need a PrintConfig

Ah, this totally makes sense.

Reply:
so should I make the change or leave it? I'll address the other comments regardless

Reply:
I think we can leave and reconsider in later PRs if PrintConfig starts to be used more liberally. Now that I think about it more, I suspect this won't happen due to the design of our Python API.